_posts/2019-03-30-numba-stencil.md

```diff
@@ -0,0 +1,323 @@
+---
+layout: post
+title: Composing Dask Array with Numba Stencils
+author: Matthew Rocklin
+tags: [dask, numba]
+theme: twitter
+---
+{% include JB/setup %}
+
+In this post we explore four array computing technologies, and how they
+work together to achieve powerful results.
+
+1.  Numba's stencil decorator to craft localized compute kernels
+2.  Numba's Just-In-Time (JIT) compiler for array computing in Python
+3.  Dask Array for parallelizing array computations across many chunks
+4.  NumPy's Generalized Universal Functions (gufuncs) to tie everything
+    together
+
+In the end we'll show how a novice developer can write a small amount of Python
+to efficiently compute localized computation on large amounts of data.  In
+particular we'll write a simple function to smooth images and apply that in
+parallel across a large stack of images.
+
+Here is the full code, we'll dive into it piece by piece below.
+
+```python
+import numba
+
+@numba.stencil
+def _smooth(x):
+    return (x[-1, -1] + x[-1, 0] + x[-1, 1] +
+            x[ 0, -1] + x[ 0, 0] + x[ 0, 1] +
+            x[ 1, -1] + x[ 1, 0] + x[ 1, 1]) // 9
```

Comment:
Would call this `mean_filter_2d` or similar. People in the image processing community will definitely know what this is. For everyone else they can easily search for more info with that name in hand (though a link to Wikipedia would be good too).

Reply:
I think I'm going to target folks other than imaging scientists to start out.  For the broader population I think that `smooth` is probably more approachable and less jargon-y.  I think that it would be good though to follow up on this blogpost with an actual imaging problem.

Maybe this is something that you would be interested?  Then maybe you and @pentschev could follow that with a GPU'ified version?

Reply:
While I agree with @mrocklin's argument about targeting a broader audience, I think it's worth a mention in the text about @jakirkham's suggestion, something like "the smooth function above/below is commonly referred to as the 2D mean filter in the image processing community".

Reply:
Where would you recommend this note?  I briefly tried looking for a place but it always seemed to break the flow of the document.

Reply:
Sounds like this was resolved by Peter's suggestion.